src/classes/CW_AudienceCSVUploadControllerTest.cls

```diff
@@ -0,0 +1,397 @@
+								where Campaign__c = :camp.Id and Name = :group2]);
+		
+	}
+	
+	static testMethod void parseFaultyMembers()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+		insert camp;
+		CW_Schedule__c sch = TestDataUtil.createSchedule(camp);
+		insert sch;
+		
+		List<Account> accs = new List<Account>();
+		for (Integer i=0;i<5;i++)
+		{
+			Account acc = TestDataUtil.createPersonAccount();
+			acc.Account_Name__pc = 'TestAccountName' + i;
+			accs.add(acc);
+		}
+		insert accs;
+		
+		
+		CW_Audience_Group__c audGr1 = TestDataUtil.createAudienceGroup(camp.Id);
+		audGr1.Name = group1;
+		audGr1.Is_Control_Group__c = false;
+		insert audGr1;
+		
+		CW_Audience_Group_Schedule__c ags = [SELECT Id, Audience_Group__c 
+											 FROM CW_Audience_Group_Schedule__c 
+											 WHERE Audience_Group__c=: audGr1.Id
+											 AND Schedule__c =: sch.Id];
+		
+		String memebrsCSVString = 'Group Name,Account Name' + CSVReader.lineTerminator;
+		
+		// create existing member for account 1
+		insert new CW_Audience_Group_Member__c(Audience_Group_LU__c = audGr1.Id, 
+					Contact__r = new Contact(Account_Name__c = accs[0].Account_Name__pc),
+					Audience_Group_Schedule__c = ags.Id);
+					
+		// add duplicate to CSV for account 2
+		memebrsCSVString += audGr1.Name + ',' + accs[1].Account_Name__pc + CSVReader.lineTerminator;
+		
+		Test.setCurrentPage(Page.CW_AudienceCSVUpload);
+		Apexpages.currentPage().getParameters().put(ConstantUtil.PAGEPARAM_CAMPAIGNID,camp.Id);
+		
+		Test.startTest();
+		CW_AudienceCSVUploadController uc = new CW_AudienceCSVUploadController();
+		uc.selectedSchId = sch.Id;
+		uc.csvFile = Blob.valueOf(memebrsCSVString);
+		uc.filename='Somefile.csv';
+		uc.uploadCSV();
+		uc.loadBatchJobProgressPage();
+		Test.stopTest();
+		
+		system.assertEquals(1,[select count() from CW_Audience_Group_Member__c where Audience_Group_LU__c = :audGr1.Id]);
+		// assert there are no duplicates
+		system.assertEquals(1,[select count() from CW_Audience_Group_Member__c 
+			where Contact__r.Account_Name__c = :accs.get(0).Account_Name__pc]);
+	}
+	
+	static testMethod void parseLFLineTerminator()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+		insert camp;
+		CW_Schedule__c sch = TestDataUtil.createSchedule(camp);
+		insert sch;
+		
+		Test.setCurrentPage(Page.CW_AudienceCSVUpload);
+		Apexpages.currentPage().getParameters().put(ConstantUtil.PAGEPARAM_CAMPAIGNID,camp.Id);
+		
+		system.assertNotEquals(ConstantUtil.LF,CSVReader.lineTerminator);
+		String stringLFCSV = groupCSVString.replace(CSVReader.lineTerminator,ConstantUtil.LF);
+		
+		Test.startTest();
+		CW_AudienceCSVUploadController uc = new CW_AudienceCSVUploadController();
+		uc.selectedSchId = sch.Id;
+		uc.csvFile = Blob.valueOf(stringLFCSV);
+		uc.filename='Somefile.csv';
+		uc.uploadCSV();
+		Test.stopTest();
+		
+		system.assertEquals(5,[select count() from CW_Audience_Group__c where Campaign__c = :camp.Id]);
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.